File: src/main/java/icu/samnyan/aqua/sega/aimedb/AimeDbDecoder.java
package icu.samnyan.aqua.sega.aimedb;

import icu.samnyan.aqua.sega.aimedb.exception.InvalidRequestException;
import icu.samnyan.aqua.sega.aimedb.util.Encryption;
import io.netty.buffer.ByteBuf;
import io.netty.channel.ChannelHandlerContext;
import io.netty.handler.codec.ByteToMessageDecoder;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.config.BeanDefinition;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;

import javax.crypto.BadPaddingException;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * @author samnyan (privateamusement@protonmail.com)
 */
@Component
@Scope(BeanDefinition.SCOPE_PROTOTYPE)
public class AimeDbDecoder extends ByteToMessageDecoder {

    private static final Logger logger = LoggerFactory.getLogger(AimeDbDecoder.class);

    private int length = 0;

    /**
     * Decrypt the incoming request including frame management
     *
     * @param ctx
     * @param in
     * @param out
     * @throws Exception
     */
    @Override
    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
        if (in.readableBytes() < 16) {
            return;
        }
        if (length == 0) {
            length = getLength(in);
        }

        if (in.readableBytes() < length) {
            return;
        }

        // Create a byte array to store the encrypted data
        ByteBuf result = Encryption.decrypt(in.readBytes(length));

        Map<String, Object> resultMap = new HashMap<>();
        resultMap.put("type", ((Short) result.getShortLE(0x04)).intValue());
        resultMap.put("data", result);

        logger.debug("Aime Server Request Type: " + resultMap.get("type"));

        out.add(resultMap);
    }

    /**
     * Get the length from request
     *
     * @param in the request
     * @return int the length of this request
     * @throws InvalidRequestException
     * @throws IllegalBlockSizeException
     * @throws InvalidKeyException
     * @throws BadPaddingException
     * @throws NoSuchAlgorithmException
     * @throws NoSuchPaddingException
     */
    private int getLength(ByteBuf in) throws InvalidRequestException, IllegalBlockSizeException, InvalidKeyException, BadPaddingException, NoSuchAlgorithmException, NoSuchPaddingException {
        int currentPos = in.readerIndex();
        ByteBuf result = Encryption.decrypt(in);
        // Check the header
        if (result.getByte(0) != 0x3e) {
            throw new InvalidRequestException();
        }

        // Read the length from offset 6

        return result.getShortLE(currentPos + 6);
    }
}

File: src/main/java/icu/samnyan/aqua/sega/aimedb/AimeDbDecoder.kt
package icu.samnyan.aqua.sega.aimedb

import io.netty.buffer.ByteBuf
import io.netty.channel.ChannelHandlerContext
import io.netty.handler.codec.ByteToMessageDecoder
import org.slf4j.Logger
import org.slf4j.LoggerFactory

/**
 * A new decoder object will be created each time a new request comes in
 */
class AimeDbDecoder : ByteToMessageDecoder() {
    companion object {
        val logger: Logger = LoggerFactory.getLogger(AimeDbDecoder::class.java)
    }

    var length = 0

    /**
     * Decrypt the incoming request including frame management
     * @param ctx ChannelHandlerContext
     * @param input ByteBuf in
     * @param out List<Object>
     */
    override fun decode(ctx: ChannelHandlerContext, input: ByteBuf, out: MutableList<Any>) {
        if (input.readableBytes() < 16) return
        if (length == 0) length = getLength(input)
        if (length < 0 || input.readableBytes() < length) return

        // Create a byte array to store the encrypted data
        val result = AimeDbEncryption.decrypt(input.readBytes(length))

        val resultMap: MutableMap<String, Any> = HashMap()
        resultMap["type"] = result.getShortLE(0x04).toInt()
        resultMap["data"] = result

        logger.debug("AimeDB Request Type: " + resultMap["type"])

        out.add(resultMap)
    }

    /**
     * Get the length from request
     *
     * @param input the request
     * @return int the length of this request
     */
    private fun getLength(input: ByteBuf): Int = try {
        val currentPos = input.readerIndex()
        val result = AimeDbEncryption.decrypt(input)

        // Check the header
        val header = result.getByte(0).toInt()
        assert(header == 0x3e) { "AimeDB: Invalid header $header" }

        // Read the length from offset 6
        result.getShortLE(currentPos + 6).toInt()
    } catch (e: Exception) {
        logger.info("AimeDB: Invalid request received")
        logger.debug("AimeDB: Invalid request received: ${input.toString(Charsets.UTF_8)}")
        -1
    }
}
